piker/brokers/kraken.py

```diff
@@ -162,6 +222,129 @@ async def _public(
         )
         return resproc(resp, log)
 
+    async def _private(
+        self,
+        method: str,
+        data: dict,
+        uri_path: str
+    ) -> Dict[str, Any]:
+        headers = {
+            'Content-Type': 
+                'application/x-www-form-urlencoded',
+            'API-Key':
+                self._api_key,
+            'API-Sign':
+                get_kraken_signature(uri_path, data, self._secret)
+        }
+        resp = await self._sesh.post(
+            path=f'/private/{method}',
+            data=data,
+            headers=headers,
+            timeout=float('inf')
+        )
+        return resproc(resp, log)
+
+    async def kraken_endpoint(
+        self,
+        method: str,
+        data: Dict[str, Any]
+    ) -> Dict[str, Any]:
+        uri_path = f'/0/private/{method}'
+        data['nonce'] = str(int(1000*time.time()))
+        resp = await self._private(method, data, uri_path)
+        return resp
+
+    async def get_positions(
+        self,
+        data: Dict[str, Any] = {}
+    ) -> Dict[str, Any]:
+        data['ofs'] = 0
+        positions = {}
+        vols = {}
+        # Grab all trade history
+        while True:
+            resp = await self.kraken_endpoint('TradesHistory', data)
+            # grab the first 50 trades
+            if data['ofs'] == 0:
+                trades = resp['result']['trades']
+            # load the next 50 trades using dict constructor
+            # for speed
+            elif data['ofs'] == 50:
+                trades = dict(trades, **resp['result']['trades'])
+            # catch the end of the trades
+            elif resp['result']['trades'] == {}:
+                count = resp['result']['count']
+                break
+            # update existing dict if num trades exceeds 100
+            else: 
+                trades.update(resp['result']['trades'])
+            # increment the offset counter
+            data['ofs'] += 50
+            # To avoid exceeding API rate limit in case of a lot of trades
+            time.sleep(1)
+
+        # make sure you grabbed all the trades
+        assert count == len(trades.values())
+
+        # positions
+        ## TODO: Make sure to add option to include fees in positions calc
+        for trade in trades.values():
+            sign = -1 if trade['type'] == 'sell' else 1
+            try:
+                positions[trade['pair']] += sign * float(trade['cost'])
+                vols[trade['pair']] += sign * float(trade['vol'])
+            except KeyError:
+                positions[trade['pair']] = sign * float(trade['cost'])
+                vols[trade['pair']] = sign * float(trade['vol'])
+    
+        for pair in positions.keys():
+            asset_balance = vols[pair]
+            if asset_balance == 0:
+                positions[pair] = 0
+            else:
+                positions[pair] /= asset_balance
```

Comment:
describe in comment what this is computing.

Reply:
See above pretty sure we can drop this loop.

Reply:
got a working version going, there is a second loop, but that is simply to package the positions in the msg format.